src/hdmf/validate/validator.py

```diff
@@ -425,55 +426,88 @@ def __init__(self, **kwargs):
             returns='a list of Errors', rtype=list)
     def validate(self, **kwargs):  # noqa: C901
         builder = getargs('builder', kwargs)
-        ret = super().validate(builder)
-        # get the data_types
-        data_types = dict()
-        for key, value in builder.items():
+        errors = super().validate(builder)
+
+        grouped_children = self.__group_children_by_data_type(builder)
+        for dt, inc_spec in self.__include_dts.items():
+            errors.extend(self.__validate_child_data_type(builder, dt, inc_spec, grouped_children))
+
+        it = chain(self.__dataset_validators.items(),
+                   self.__group_validators.items())
+        for name, validator in it:
+            errors.extend(self.__validate_untyped_child(builder, name, validator))
+
+        return errors
+
+    def __group_children_by_data_type(self, builder):
+        """Returns a map of the children of a builder grouping those children by data_type
+
+        The keys of the map are data_type names and they map to lists of all child builders of that data_type
+        """
+        data_types = defaultdict(list)
+        for _, value in builder.items():
             v_builder = value
```

Comment:
```suggestion
        for v_builder in builder.values():
```

Reply:
Good suggestion to just iterate over `.values()`, I included that change.

Removing the assignment on line 449 also looks reasonable, but on closer inspection, it would end up changing the logic for one particular edge case: if `value` is a `LinkBuilder` which links to something inheriting from `BaseBuilder`, and that `BaseBuilder` has a `data_type`, then it would store the `BaseBuilder` in the `data_types` map, as opposed to the current behavior which is to store the `LinkBuilder` in the map.

excerpt for reference:
```python
            v_builder = value
            if isinstance(v_builder, LinkBuilder):
                v_builder = v_builder.builder
            if isinstance(v_builder, BaseBuilder):
                dt = v_builder.attributes.get(self.spec.type_key())
                if dt is not None:
                    data_types[dt].append(value)
```

The implementation here hasn't changed from the previous version, it was just extracted into a separate function. Looking at the usage of the return value of `__group_children_by_data_type` after it is passed to `__validate_child_data_type`, there is another check if the builder is a `LinkedBuilder` (line 467), so it seems to me like the current implementation is consistent. However, by making the change you suggested, all unit tests still pass, so it's not obvious that making that change would be incorrect.

I think there will likely be some large changes to validation logic in the solution for the new issue: https://github.com/hdmf-dev/hdmf/issues/507 so my suggestion is to verify that the solution for that issue verifies correct behavior here. Do you agree?

Reply:
You are right. How you have it now, keeping the assignment on line 449, is correct. We want to pass on the fact that a `LinkBuilder` is associated with the data type. Good catch.